src/main/scala/it/pagopa/interop/backendforfrontend/api/impl/AgreementsApiServiceImpl.scala

```diff
@@ -41,6 +73,113 @@ final case class AgreementsApiServiceImpl(agreementProcessService: AgreementProc
+    }
+  }
+
+  def enhanceAgreement(
+    agreement: AgreementProcess.Agreement
+  )(implicit contexts: Seq[(String, String)]): Future[Agreement] = for {
+    producer          <- partyProcessService.getInstitution(agreement.producerId)
+    consumer          <- partyProcessService.getInstitution(agreement.consumerId)
+    consumerTenant    <- tenantManagementService.getTenant(agreement.consumerId)
+    eService          <- catalogManagementService.getEService(agreement.eserviceId)
+    currentDescriptor <- eService.descriptors
+      .find(_.id == agreement.descriptorId)
+      .toFuture(AgreementDescriptorNotFound(agreement.id))
+    activeDescriptor = eService.descriptors.sortBy(_.version.toInt).lastOption
+
+    allAttributesIds = (eServiceAttributesIds(eService) ++ tenantAttributesIds(consumerTenant)).distinct
+    attributes <- attributeRegistryService.getBulkAttributes(allAttributesIds)
+
+    agreementVerifiedAttrs  = filterAttributes(attributes, agreement.verifiedAttributes.map(_.id))
+      .map(_.toVerifiedAttribute)
+    agreementCertifiedAttrs = filterAttributes(attributes, agreement.certifiedAttributes.map(_.id))
+      .map(_.toCertifiedAttribute)
+    agreementDeclaredAttrs  = filterAttributes(attributes, agreement.declaredAttributes.map(_.id))
+      .map(_.toDeclaredAttribute)
+
+    tenantAttributes = consumerTenant.attributes.map(enhanceTenantAttribute(attributes, _))
+  } yield Agreement(
+    id = agreement.id,
+    descriptorId = agreement.descriptorId,
+    producer = Tenant(id = agreement.producerId, name = producer.description),
+    consumer =
+      TenantWithAttributes(id = agreement.consumerId, name = consumer.description, attributes = tenantAttributes),
+    eservice = EService(
+      id = agreement.eserviceId,
+      name = eService.name,
+      version = currentDescriptor.version,
+      activeDescriptor = activeDescriptor.map(_.toActiveDescriptor)
+    ),
+    state = agreement.state.toApi,
+    verifiedAttributes = agreementVerifiedAttrs,
+    certifiedAttributes = agreementCertifiedAttrs,
+    declaredAttributes = agreementDeclaredAttrs,
+    suspendedByConsumer = agreement.suspendedByConsumer,
+    suspendedByProducer = agreement.suspendedByProducer,
+    suspendedByPlatform = agreement.suspendedByPlatform,
+    consumerNotes = agreement.consumerNotes,
+    consumerDocuments = agreement.consumerDocuments.map(_.toApi),
+    createdAt = agreement.createdAt,
+    updatedAt = agreement.updatedAt
+  )
+
+  def eServiceAttributesIds(eService: CatalogManagement.EService): Seq[UUID] =
+    eService.attributes.verified.flatMap(_.single.map(_.id)) ++
+      eService.attributes.verified.flatMap(_.group).flatten.map(_.id) ++
+      eService.attributes.declared.flatMap(_.single.map(_.id)) ++
+      eService.attributes.declared.flatMap(_.group).flatten.map(_.id) ++
+      eService.attributes.certified.flatMap(_.single.map(_.id)) ++
+      eService.attributes.certified.flatMap(_.group).flatten.map(_.id)
```

Comment:
How about this? It requires `cats.implicits` but uses `.mapFilter` that is STRICTLY `Traversable[X].mapFilter(_ => Option[X]): Traversable[X]` and doesn't rely on the fact that `Seq` and `Option` have a LUB that is `TraversableOnce[X]`.
It's a bit more verbose but possibly more clear.

```suggestion
  def eServiceAttributesIds(eService: CatalogManagement.EService): Seq[UUID] = {
    val attrs: Seq[CatalogManagement.Attribute] =
      eService.attributes.verified ++ eService.attributes.declared ++ eService.attributes.certified

    attrs
      .mapFilter(a =>
        (a.single, a.group) match {
          case (Some(s), Some(g)) => Some(s :: g.toList)
          case (None, x)          => x
          case (Some(s), None)    => Some(s :: Nil)
          case (None, None)       => None
        }
      )
      .flatten
      .map(_.id)
  }
```

Reply:
❤️ 
(I just removed the last case because is already handled by the second)